File: data/derived_cot/rq3_prompting/results/translation/glm-5.1/py_to_cpp/code_output/CombinationCalculator.cpp
#include <vector>
#include <string>
#include <cmath>

class CombinationCalculator {
public:
    std::vector<std::string> datas;

    CombinationCalculator(const std::vector<std::string>& datas) : datas(datas) {}

    static long long count(int n, int m) {
        if (m == 0 || n == m) return 1;
        if (m > n - m) m = n - m;
        long long result = 1;
        for (int i = 0; i < m; i++) {
            result = result * (n - i) / (i + 1);
        }
        return result;
    }

    static double count_all(int n) {
        if (n < 0 || n > 63) return 0;
        if (n == 63) return INFINITY;
        return (1LL << n) - 1;
    }

    std::vector<std::vector<std::string>> select(int m) {
        std::vector<std::vector<std::string>> result;
        std::vector<std::string> resultList(m);
        _select(0, resultList, 0, result);
        return result;
    }

    std::vector<std::vector<std::string>> select_all() {
        std::vector<std::vector<std::string>> result;
        for (int i = 1; i <= static_cast<int>(datas.size()); i++) {
            auto sub = select(i);
            result.insert(result.end(), sub.begin(), sub.end());
        }
        return result;
    }

    void _select(int dataIndex, std::vector<std::string>& resultList, int resultIndex, std::vector<std::vector<std::string>>& result) {
        int resultLen = static_cast<int>(resultList.size());
        int resultCount = resultIndex + 1;
        if (resultCount > resultLen) {
            result.push_back(resultList);
            return;
        }
        int loopEnd = static_cast<int>(datas.size()) + resultCount - resultLen;
        for (int i = dataIndex; i < loopEnd; i++) {
            resultList[resultIndex] = datas[i];
            _select(i + 1, resultList, resultIndex + 1, result);
        }
    }
};

File: data/derived_cot/rq2_eval/translation/r1/_converted_code/java_to_py/CombinationCalculator.py
import math

class CombinationCalculator:

    def __init__(self, datas):
        self.datas = datas

    @staticmethod
    def count(n, m):
        if m == 0 or n == m:
            return 1
        return CombinationCalculator.factorial(n) // (CombinationCalculator.factorial(n - m) * CombinationCalculator.factorial(m))

    @staticmethod
    def factorial(x):
        result = 1
        for i in range(1, x + 1):
            result *= i
        return result

    @staticmethod
    def count_all(n):
        if n < 0 or n > 63:
            return math.nan
        if n == 63:
            return math.inf
        n_mod = n % 32
        return float((1 << n_mod) - 1)

    def select(self, m):
        result = []
        self._select(0, [], 0, result, m)
        return result

    def select_all(self):
        result = []
        n = len(self.datas)
        for i in range(1, n + 1):
            result.extend(self.select(i))
        return result

    def _select(self, dataIndex, resultList, resultIndex, result, m):
        if resultIndex == m:
            result.append(resultList[:])
            return
        n = len(self.datas)
        for i in range(dataIndex, n - (m - resultIndex) + 1):
            resultList.insert(resultIndex, self.datas[i])
            self._select(i + 1, resultList, resultIndex + 1, result, m)
            resultList.pop(resultIndex)